### PR TITLE
Recaptcha

### PR DESCRIPTION
**Pull request by gastonsoto**
Adding recaptcha. Default to V3, fallback to V2 based on v3 score. Implemented on Login, Register, Contact. Keys added using Rails encrypted credentials:

EDITOR="code --wait" bin/rails credentials:edit --environment [environment here]

Note: code is the editor you use, I'm using vscode right now.

Reference:

https://api.rubyonrails.org/classes/Rails/Application.html#method-i-credentials

### CODE REVIEW

**Comment by gastonsoto** [comment]
Checking the test, everything went well locally.

File: app/controllers/concerns/recaptchable.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  extend ActiveSupport::Concern
+
+  included do
+    def verify_recaptchas
+      v2_secret_key = Rails.application.credentials.recaptcha_v2[:secret_key]
+      v3_secret_key = Rails.application.credentials.recaptcha_v3[:secret_key]
+
+      success = verify_recaptcha(
+        action: params[:action],
+        minimum_score: 0.5,
```
Comment (martinlevesque) [inline app/controllers/concerns/recaptchable.rb]: 0.5 comes from what/where?

**Comment by gastonsoto** [comment]
> Checking the test, everything went well locally.

Adding encrypted keys to travis CI.

File: app/controllers/concerns/recaptchable.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        secret_key: v3_secret_key
+      )
+
+      unless success
+        checkbox_success = verify_recaptcha(secret_key: v2_secret_key)
+      end
+
+      @show_checkbox_recaptcha = !success
```
Comment (martinlevesque) [inline app/controllers/concerns/recaptchable.rb]: maybe rename to something like @show_v2_recaptcha ? just a suggestion

Comment (gastonsoto) [reply app/controllers/concerns/recaptchable.rb]: https://developers.google.com/recaptcha/docs/v3

Scroll down to "Interpreting the score"

File: app/views/home/support.html.erb [hunk 3]
```diff
@@ -1,5 +1,47 @@
...
+            <%= text_area_tag :message, '', class:'form-control' %>
+          </div>
+
+          <div class="form-label-group mb-3">            
+            <%= check_box_tag :newsletter, '', class:'form-control' %>
+            <%= label_tag :newsletter, 'Do you want to receive our newsletter?' %>
+          </div>    
+
+          <div class="form-label-group mb-3">
+            <%= render_recaptcha('support',@show_checkbox_recaptcha) %>
```
Comment (martinlevesque) [inline app/views/home/support.html.erb]: add a space, render_recaptcha('support', @show_checkbox_recaptcha)

File: app/controllers/concerns/recaptchable.rb (hunk 2 above)
Comment (gastonsoto) [reply app/controllers/concerns/recaptchable.rb]: Oh shit, this was for the other question.

> 0.5 comes from what/where?

File: app/views/users/forgot_password.html.erb [hunk 4]
```diff
@@ -13,7 +13,11 @@
             <%= label_tag :email, 'Email', placeholder: 'your@email.com' %>
             <%= text_field_tag :email, '', class:'form-control' %>
           </div>                   
-
+          
+          <div class="form-label-group mb-3">
+            <%= render_recaptcha('forgot_password',@show_checkbox_recaptcha) %>
```
Comment (martinlevesque) [inline app/views/users/forgot_password.html.erb]: render_recaptcha('forgot_password', @show_checkbox_recaptcha)

File: app/controllers/concerns/recaptchable.rb (hunk 2 above)
Comment (gastonsoto) [reply app/controllers/concerns/recaptchable.rb]: > maybe rename to something like @show_v2_recaptcha ? just a suggestion

yes that works

File: app/controllers/concerns/recaptchable.rb (hunk 1 above)
Comment (martinlevesque) [reply app/controllers/concerns/recaptchable.rb]: looks good

File: app/views/home/support.html.erb [hunk 5]
```diff
@@ -1,5 +1,47 @@
-<section class="support-header bg-light px-3 py-3 pt-md-5 pb-md-4 mx-auto">
+<section class="public-support-header bg-light px-3 py-3 pt-md-5 pb-md-4 mx-auto">
```
Comment (martinlevesque) [inline app/views/home/support.html.erb]: - Remove the newsletter checkbox
- How can we help you? -> Increase the size of this box

image

File: app/views/users/register.html.erb [hunk 6]
```diff
@@ -23,11 +23,14 @@
             <%= text_field_tag :password_confirmation, '', class:'form-control' %>
           </div>  
 
-          <div class="form-label-group mb-3">
-            <%= label_tag :newsletter %>
+          <div class="form-label-group mb-3">
```
Comment (martinlevesque) [inline app/views/users/register.html.erb]: text fields should be required

File: app/views/layouts/_navbar.html.erb [hunk 7]
```diff
@@ -55,7 +80,7 @@
           <%= link_to "Log In", login_path, class:'nav-link' %>
```
Comment (martinlevesque) [inline app/views/layouts/_navbar.html.erb]: Unable to login due to recaptcha

Comment (gastonsoto) [reply app/views/layouts/_navbar.html.erb]: Fixed this, the wrong method was being called in sessions controller. BUT if v3 has issues to validate, then the page renders again with V2 checkbox. If this becomes an issue, you can just render v2 directly by passing `show_recaptcha_v2:true`:

`<%= render_recaptcha('register', true %>`

File: app/controllers/sessions_controller.rb [hunk 8]
```diff
@@ -8,9 +10,12 @@ def create
 
     token = api(:post, '/account/getToken', payload: payload)
 
-    session[:token] = token
-
-    redirect_to root_url, notice: "Logged in!"
+    if verify_recaptchas && token.present?
```
Comment (martinlevesque) [inline app/controllers/sessions_controller.rb]: On login the recaptcha v3 fails

image

File: app/views/layouts/_navbar.html.erb (hunk 7 above)
Comment (martinlevesque) [reply app/views/layouts/_navbar.html.erb]: v3 does not work on login on my local, normal?

File: app/views/layouts/_navbar.html.erb [hunk 9]
```diff
@@ -55,7 +80,7 @@
...
         </li>
         <li class="nav-item">
-          <%= link_to "Sign Up", users_register_path, class:'btn btn-sn btn-outline-secondary' %>
+          <%= link_to "Sign Up", users_register_path, class:'btn btn-sn btn-outline-secondary hvr-fade' %>
```
Comment (martinlevesque) [inline app/views/layouts/_navbar.html.erb]: always put a space after ":", 

class:'btn btn-sn btn-outline-secondary hvr-fade' -> class: 'btn btn-sn btn-outline-secondary hvr-fade'

everywhere

File: app/views/layouts/_password_field.html.erb [hunk 10]
```diff
@@ -0,0 +1,9 @@
+<%= label_tag name.to_sym, label, class:'required' %>
+<div class="input-group show_hide_password">             
+  <%= password_field_tag name.to_sym, '', class:'form-control', required:true %>
```
Comment (martinlevesque) [inline app/views/layouts/_password_field.html.erb]: spaces after :

**Comment by gastonsoto** [comment]
Updated.

File: app/views/users/register.html.erb [hunk 11]
```diff
@@ -5,31 +5,32 @@
...
 <section class="pt-8 pt-md-11 bg-dark p-3 p-md-5">
   <div class="container">
-    <%= form_tag users_path, {class:'form-signin'} do |form| %>
+    <%= form_for :account, {url:users_path, html: {class: 'form-signin'}} do |form| %>
       <div class="card">
         <div class="card-body">
           <div class="form-label-group mb-2">
-            <%= label_tag :email %>
-            <%= text_field_tag :email, '', class:'form-control' %>
+            <%= form.label 'Email', class: 'required' %>
```
Comment (martinlevesque) [inline app/views/users/register.html.erb]: when you click on label email, it should focus the email field

File: app/views/users/register.html.erb [hunk 12]
```diff
@@ -5,31 +5,32 @@
...
-            <%= text_field_tag :password_confirmation, '', class:'form-control' %>
+            <%= render 'layouts/password_field', name: 'account[password_confirmation]', label: 'Confirm your password' %>
           </div>  
 
-          <div class="form-label-group mb-3">
-            <%= label_tag :newsletter %>
-            <%= check_box_tag :newsletter, '', class:'form-control' %>
+          <div class="form-label-group mb-3">            
+            <%= form.check_box :newsletter, { checked: true }, 1, 0 %>
+            <%= form.label 'Do you want to receive our newsletter?' %>
```
Comment (martinlevesque) [inline app/views/users/register.html.erb]: when you click on "Do you want to receive ..." it should check-uncheck

**Pull request merged**